test/core/networking/websockets/middleware/encryption_middleware_test.dart

```diff
@@ -0,0 +1,229 @@
+import 'dart:io';
+
+import 'package:test/test.dart';
+
+import 'package:flyde/core/networking/websockets/middleware/middleware_types.dart';
+import 'package:flyde/core/networking/websockets/middleware/encryption_middleware.dart';
+import 'package:flyde/core/networking/websockets/session.dart';
+
+import '../../../../helpers/value_hook.dart';
+import '../../../../helpers/get_uri.dart';
+import '../../../../helpers/open_webserver.dart';
+
+void main() {
+  test('EncryptionMiddleware can handshake', () async {
+    final receivedMessage = VHook<String?>(null);
+
+    final List<MiddlewareFunc> middleware = [encryptionMiddleware];
+
+    // Create server to accept connection
+    final server = await openWebServer();
+    server.wsMiddleware = middleware;
+    server.wsOnMessage = (ServerSession session, dynamic msg) async {
+      receivedMessage.set(msg);
+    };
+
+    // Connect to the server
+    final session = ClientSession(getUri(server, 'ws').toString());
+    session.middleware = middleware;
+
+    session.send('ANYTHING');
+    // Check message transmitted
+    await receivedMessage.awaitValue(Duration(seconds: 10), raiseOnTimeout: true);
+
+    receivedMessage.expect(equals("ANYTHING"));
+    expect(session.storage['crypto_provider'], equals(isA<CryptoProvider>()));
+  });
+
+  test('EncryptionMiddleware throws on crypto message after shared key is established', () async {
+    final receivedNormalMessage = VHook<String?>(null);
+    final receivedCryptoMessage = VHook<String?>(null);
+    final raisedException = VHook<Object?>(null);
+    final closedSession = VHook<bool?>(null);
+
+    final List<MiddlewareFunc> middleware = [encryptionMiddleware];
+
+    // Create server to accept connection
+    final server = await openWebServer();
+    server.wsMiddleware = middleware;
+    server.wsOnMessage = (ServerSession session, dynamic msg) async {
+      if (msg == 'ANYTHING-1') {
+        receivedNormalMessage.set(msg);
+      } else {
+        receivedCryptoMessage.set(msg);
+      }
+    };
+    server.wsOnError = (ServerSession session, Object exception) {
+      raisedException.set(exception);
+    };
+
+    // Connect to the server
+    final session = ClientSession(getUri(server, 'ws').toString());
+    session.middleware = middleware;
+    session.onDone = (dynamic session) {
+      closedSession.set(true);
+    };
+
+    session.send('ANYTHING-1');
+    // Check message transmitted (extra timeout for key generation)
+    await receivedNormalMessage.awaitValue(Duration(seconds: 10), raiseOnTimeout: true);
+
+    // Send crypto message
+    session.send('\$ANYTHING-2');
```

Comment:
```suggestion
    session.send(r'$ANYTHING-2');
```

Should be changed at a few more spots